### PR TITLE
Refactor Sholl analysis code, fix #663

### LINKED ISSUE #663: Sholl analysis is performing a lot of unnecessary computations
The results of Sholl analysis return of a lot of zeros. I _think_ this is because the bounding box encloses the whole neuron while it should enclose only the trees of a specific neurite_type on which the computation is actually performed. See the following line:

https://github.com/BlueBrain/NeuroM/blob/49dbaef7ed78c351d8adb3085843010716afa217/neurom/fst/_neuronfunc.py#L201

### PR DESCRIPTION
**Pull request by asanin-epfl**
Neuron attributes soma center and soma radius were being computed for each neurite. Now they are computed only once.

### CODE REVIEW

File: neurom/features/neuronfunc.py [hunk 1]
```diff
@@ -244,18 +244,18 @@ def _count_crossings(neurite, radius):
...
 @feature(shape=(...,))
-def sholl_frequency(nrn, neurite_type=NeuriteType.all, step_size=10):
+def sholl_frequency(nrn, neurite_type=NeuriteType.all, bins=None):
     """Perform Sholl frequency calculations on a population of neurites.
 
     Args:
         nrn(morph): nrn or population
         neurite_type(NeuriteType): which neurites to operate on
-        step_size(float): step size between Sholl radii
+        bins(iterable of floats|int|None): binning to use for the Sholl radii
```
Comment (asanin-epfl) [inline neurom/features/neuronfunc.py:258]: @arnaudon can you please add more documentation to this parameter?

File: neurom/features/neuronfunc.py [hunk 2]
```diff
@@ -244,18 +244,20 @@ def _count_crossings(neurite, radius):
...
 
     return np.array([sum(_count_crossings(neurite, r)
-                         for neurite in iter_neurites(neurites))
+                         for neurite in iter_neurites(neurites, filt=neurite_filter))
                      for r in radii])
 
 
 @feature(shape=(...,))
-def sholl_frequency(nrn, neurite_type=NeuriteType.all, step_size=10):
+def sholl_frequency(nrn, neurite_type=NeuriteType.all, bins=None):
```
Comment (arnaudon) [inline neurom/features/neuronfunc.py:252]: why not bins=10 here, instead of if None below?

File: neurom/features/neuronfunc.py (hunk 1 above)
Comment (arnaudon) [reply neurom/features/neuronfunc.py]: ah sorry, I just saw you did it, it's good to me, just the thing with bins=None I find a bit weird, but as you prefer

Comment (asanin-epfl) [reply neurom/features/neuronfunc.py]: done

**Review by arnaudon — APPROVED** [review]
LGTM!

**Comment by asanin-epfl** [comment]
@GianlucaFicarelli the PR has been reviewed by Alexis. I will need your approval to merge.

**Comment by arnaudon** [comment]
can't you give me and adrien some more access so we can approve witout merge?

**Review by GianlucaFicarelli — APPROVED** [review]

**Comment by asanin-epfl** [comment]
Thank you!

File: tests/features/test_get_features.py [hunk 3]
```diff
@@ -859,19 +859,19 @@ def test_soma_surface_areas():
 
 def test_sholl_frequency():
     assert_allclose(get_feature('sholl_frequency', NEURON),
-                    [4, 8, 8, 14, 9, 8, 7, 7])
+                    [4, 6, 10, 8, 8, 11, 7, 9, 8, 8])
```
Comment (mgeplf) [inline tests/features/test_get_features.py:862]: how were these checked to be correct?

I notice in https://github.com/BlueBrain/NeuroM/issues/663 I asked for a fake neuron to test this on, could we add one to make sure we're catching the edge cases?

**Comment by mgeplf** [comment]
@arnaudon > can't you give me and adrien some more access so we can approve witout merge?

Yes, that isn't right.

File: tests/features/test_get_features.py (hunk 3 above)
Comment (asanin-epfl) [reply tests/features/test_get_features.py:862]: @arnaudon may be you have additional inquiries for tests here as well?

**Comment by arnaudon** [comment]
Ah hold on! @mariarv  is just telling me that we need step_size, not bins as argument, otherwise we loose the information of the distance of each sholl sphere, I didn't realise this before, sorry... We can still provide a list of bins, but it is easier to have step_size. I'll make another PR on that, and check the tests.

File: tests/features/test_get_features.py (hunk 3 above)
Comment (arnaudon) [reply tests/features/test_get_features.py:862]: I think we can simply check by eye on the morphologies, by plotting circles for each sholl sphere and counting? not sure it is worth adding another test morph for that.

Comment (mgeplf) [reply tests/features/test_get_features.py:862]: By engineering a morphology that has all the edge-cases, we can be more sure that we're covering them.
For instance; there was debate on what happens when a section crosses back across the Sholl spheres.
